src/lib/group.ts

```diff
@@ -0,0 +1,394 @@
+import { Scalar, withMessage } from './core.js';
+import { Field, FieldVar, isField } from './field.js';
+import { Bool, Snarky } from '../snarky.js';
+import { Field as Fp } from '../provable/field-bigint.js';
+import { Pallas } from '../bindings/crypto/elliptic_curve.js';
+
+export { Group };
+
+type FieldLike = FieldVar | Field | number | string | bigint;
+type GroupLike = {
+  x: FieldLike;
+  y: FieldLike;
+};
+
+/**
+ * An element of a Group.
+ */
+class Group {
+  x: Field;
+  y: Field;
+
+  /**
+   * The generator `g` of the Group.
+   */
+  static get generator() {
+    return new Group(Pallas.one.x, Pallas.one.y);
+  }
+
+  /**
+   * Coerces anything group-like to a {@link Group}.
+   */
+  constructor(g: {
+    x: FieldVar | Field | number | string | bigint;
+    y: FieldVar | Field | number | string | bigint;
+  });
+  /**
+   * Coerces anything group-like to a {@link Group}.
+   */
+  constructor(
+    x: FieldVar | Field | number | string | bigint,
+    y: FieldVar | Field | number | string | bigint
+  );
+  constructor(
+    arg1: GroupLike | FieldLike,
+    arg2: FieldLike | undefined = undefined
+  ) {
+    let x, y: FieldLike;
+
+    if (isGroupLike(arg1)) {
+      x = arg1.x;
+      y = arg1.y;
+    } else {
+      x = arg1 as FieldLike;
+      y = arg2 as FieldLike;
+    }
+
+    this.x = isField(x) ? x : new Field(x);
+    this.y = isField(y) ? y : new Field(y);
+
+    /*
+    technically elements are only group elements if they are on the curve (aka for two points y, x the following equation must hold y^2 = x^3 + 5 over our field)
```

Comment:
I briefly talked to Gregor about this; the constructor doesn't check if a (constant) element is on the curve - it only does so if its a witness. Adding this would add constraints that could break existing contracts. That could be something for a future PR